flashflood/util.py

```diff
@@ -89,16 +89,32 @@ def _delete(keys_to_delete):
                 f.result()
 
 class S3Deleter(AbstractContextManager):
-    def __init__(self, bucket, deletion_threshold=5 * _S3_BATCH_DELETE_MAX_KEYS):
+    def __init__(self, bucket, deletion_threshold=5 * _S3_BATCH_DELETE_MAX_KEYS, number_of_workers=4):
         self.bucket = bucket
         self.deletion_threshold = deletion_threshold
         self._keys = list()
+        self.number_of_workers = number_of_workers
 
     def delete(self, key):
         self._keys.append(key)
         if self.deletion_threshold <= len(self._keys):
-            delete_keys(self._keys)
+            delete_keys(self._keys, number_of_workers=self.number_of_workers)
             self._keys = list()
 
     def __exit__(self, *args, **kwargs):
         delete_keys(self.bucket, self._keys)
+
+def concurrent_listing(bucket, prefixes, number_of_workers=4):
+    """
+    Concurrently list objects from `bucket` for `prefixes`. Lexicographical ordering is lost.
+    """
+    assert not isinstance(prefixes, str)
+
+    def _list(pfx):
+        return [item for item in bucket.objects.filter(Prefix=pfx)]
```

Comment:
Should we worry about this list growing too large?

Reply:
Probably. I'll add a TODO